### DISCUSSION TITLE
Phase4 Error responses are not signed which leads to "Unable to verify document signature" error on failed sending attempt instead of the actual error message

### DISCUSSION THREAD

**Discussion by alfchristensen** [OxalisCommunity/oxalis#626] [Q&A]
Hi!

After the successful test using the standalone to send a peppol document, I now need to implement the maven dependency in order to send documents using our actual application. I am using version 5.5.0 of both oxalis and as4.

When trying to return the test document to the testbed I get the following error:
"javax.xml.crypto.MarshalException: It is forbidden to use algorithm http://www.w3.org/2000/09/xmldsig#rsa-sha1 when secure validation is enabled" and "network.oxalis.api.lang.OxalisTransmissionException: Unable to verify document signature".

Any idea how I can solve this? Please see my pom.xml, oxalis.conf, java class, and log attached in the zip file.

Best Regards
Alf Christensen
[files.zip](https://github.com/OxalisCommunity/oxalis/files/11689627/files.zip)

**Comment by dladlk** [comment]
Please check the java version you use - such error can occur in java 17, where secure validation is activated by default.
Oxalis 5.5.0 requires java 8.

**Reply by dladlk** [reply]
BTW, it could be also some misconfiguration with AS4 module installation - because CXF, used in Oxalis-AS4, overwrites default security validation rules, and allows SHA-1, used in the current version of SMP specification.

**Reply by dladlk** [reply]
And looked into your pom.xml - you need to change 17 to 8:

```
    <properties>
        <maven.compiler.source>17</maven.compiler.source>
        <maven.compiler.target>17</maven.compiler.target>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
    </properties>
```

**Reply by alfchristensen** [reply]
Thanks for your feedback @dladlk :) I am building this function as its own app now, but it will be implemented in an application running java 17. Downgrading Java version is unfortunately not an option. If so, I would have to make a local API, and really hope to avoid that. Also, my access point with the standalone and as4 add on works on java 11. 

Misconfiguration with the AS4 module is entirely possible. Do you see anything missing or something wrong with in the oxalis.conf file? 

Thanks for the part about CFX overwriting default security values. Will look into that.

**Reply by dladlk** [reply]
Alternative can be to try to use Oxalis 6.0.0 release candidate 4 from snapshot repository, which supports java 11:
https://oss.sonatype.org/content/repositories/snapshots/network/oxalis/oxalis-as4/6.0.0-RC4-SNAPSHOT/
For this I think you need to configure snapshots repository in your pom.xml:

```
    <distributionManagement>
        <snapshotRepository>
            <id>ossrh</id>
            <url>https://oss.sonatype.org/content/repositories/snapshots</url>
        </snapshotRepository>
        <repository>
            <id>ossrh</id>
            <url>https://oss.sonatype.org/service/local/staging/deploy/maven2/</url>
        </repository>
    </distributionManagement>
```

and switch from 5.5.0 with 6.0.0-RC4-SNAPSHOT

I use it in my local tests and it works ok, but it is not yet released.

But I am surprised that your environment managed to achieve lookup phase and parsed the reply from SMP, as multiple javax packages are removed from Java 11. Looks like you included javax libraries into your classpath. It gives one more way to solve the issue - modify security.policy in your JRE and comment out forbidden SHA1 algorithms - delete lines 

disallowAlg http://www.w3.org/2000/09/xmldsig#sha1,\
and
disallowAlg http://www.w3.org/2000/09/xmldsig#rsa-sha1,\

in "jdk-17\conf\security\java.security"

**Reply by alfchristensen** [reply]
Thanks for your last reply @dladlk 

Now I have tried to use Java 8, 11, 17 with oxalis 5.0.0, 5.5.0 and 6.0.0-RC3. 
Java 17 and all 3 Oxalis versions returns the "SHA1 forbidden" error message. 
When removing the two lines you mentioned, I got the same error as for Java 8 and 11 below.

Java 8 and 11 and Oxalis 5.0.0 and 5.5.0 return the following error message:
"SEVERE: Inbound policy verification failed: These policy alternatives can not be satisfied: 
{http://docs.oasis-open.org/ws-sx/ws-securitypolicy/200702}X509Token: The received token does not match the token inclusion requirement
{http://docs.oasis-open.org/ws-sx/ws-securitypolicy/200702}SignedParts: Soap Body is not SIGNED"

I should mention that I have an Access Point running the docker image, and that works with the standalone.jar. Is there an API on the access point I can send my peppol documents to? Do I need to encrypt or sign the inputstream with an x509 certificate myself, or is it enough to parse a plain xml file to an input stream? I don't really understand the documentation on how to implement this using either the maven dependency or the API (if it exists?)

@aaron-kumar is there any possibility that you can update the signing algoritm from SHA1 to something that Java 17 will approve?

**Reply by dladlk** [reply]
SHA1 is a part of the current version of SMP specification, see https://github.com/OpenPEPPOL/documentation/blob/master/TransportInfrastructure/ICT-Transport-SMP_Service_Specification-110.pdf , chapter 5.5 Security, which says:
![image](https://github.com/OxalisCommunity/oxalis/assets/41484081/b5e73716-356d-40b7-9897-d46a807ea7c6)

So until it is upgraded to SHA256 or newer, Oxalis has to work with SHA1, which is disallowed by default in Java 17.

Regarding "Soap Body is not SIGNED" - try to activate in your logback.xml of sending Oxalis a DEBUG level for `org.apache.cxf` like adding `<logger name="org.apache.cxf" level="DEBUG"/>` to see details of the response you receive, which should include actual reason of the rejection, potentially unsigned and rejected by Oxalis...

**Reply by alfchristensen** [reply]
Thanks for your reply!
My apologies. Here is the complete log attached.
[java-11-oxalis-6.0.0-RC3.log](https://github.com/OxalisCommunity/oxalis/files/11722747/java-11-oxalis-6.0.0-RC3.log)

**Reply by dladlk** [reply]
OK - you response contains next XML:

```
<?xml version="1.0" encoding="UTF-8"?>
<S12:Envelope xmlns:S12="http://www.w3.org/2003/05/soap-envelope">
	<S12:Header>
		<eb:Messaging xmlns:S11="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ds="http://www.w3.org/2000/09/xmldsig#" xmlns:eb="http://docs.oasis-open.org/ebxml-msg/ebms/v3.0/ns/core/200704/" xmlns:ebbp="http://docs.oasis-open.org/ebxml-bp/ebbp-signals-2.0" xmlns:wsu="http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-wssecurity-utility-1.0.xsd" xmlns:xlink="http://www.w3.org/1999/xlink" wsu:Id="phase4-msg-1b3fe283-466c-4aa9-a569-04ad10f8fcea">
			<eb:SignalMessage>
				<eb:MessageInfo>
					<eb:Timestamp>2023-06-12T11:59:08.247Z</eb:Timestamp>
					<eb:MessageId>605c56d2-8639-4ac6-88f7-df018d399fa3@phase4.openpeppol.playground</eb:MessageId>
					<eb:RefToMessageId>40c87c09-6f96-4b53-8bc3-d6eda585c389@LAPTOP-HCUTN7PI</eb:RefToMessageId>
				</eb:MessageInfo>
				<eb:Error category="Content" errorCode="EBMS:0004" refToMessageInError="40c87c09-6f96-4b53-8bc3-d6eda585c389@LAPTOP-HCUTN7PI" severity="failure" shortDescription="Other">
					<eb:Description xml:lang="en">
					Invoked AS4 message processor SPI com.helger.phase4.peppol.servlet.Phase4PeppolServletMessageProcessorSPI@401cdd3 on '40c87c09-6f96-4b53-8bc3-d6eda585c389@LAPTOP-HCUTN7PI' 
					returned a failure: 
					Failed to extract the Peppol data from SBDH. Technical details: com.helger.peppol.sbdh.read.PeppolSBDHDocumentReadException - 
					The 'DocumentIdentification/Standard' element has the illegal value 
					'urn:oasis:names:specification:ubl:schema:xsd:Invoice-2'. 
					It must equal the namespace URI of the business message ('urn:oasis:names:specification:ubl:schema:xsd:Invoice-2'). 
					It must also be the beginning of the document type identifier value ('
                    urn:oasis:names:specification:ubl:schema:xsd:Invoice-2::Invoice##urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0::2.1').</eb:Description>
					<eb:ErrorDetail>An undefined error occurred.</eb:ErrorDetail>
				</eb:Error>
			</eb:SignalMessage>
		</eb:Messaging>
	</S12:Header>
	<S12:Body/>
</S12:Envelope>

```

Error message there:

```
Invoked AS4 message processor SPI com.helger.phase4.peppol.servlet.Phase4PeppolServletMessageProcessorSPI@401cdd3 on '40c87c09-6f96-4b53-8bc3-d6eda585c389@LAPTOP-HCUTN7PI' 
returned a failure: 
Failed to extract the Peppol data from SBDH. Technical details: com.helger.peppol.sbdh.read.PeppolSBDHDocumentReadException - 
The 'DocumentIdentification/Standard' element has the illegal value 
'urn:oasis:names:specification:ubl:schema:xsd:Invoice-2'. 
It must equal the namespace URI of the business message ('urn:oasis:names:specification:ubl:schema:xsd:Invoice-2'). 
It must also be the beginning of the document type identifier value ('
urn:oasis:names:specification:ubl:schema:xsd:Invoice-2::Invoice##urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0::2.1').
```

I think your SBDH should be updated. You can try to use the default SBD wrapper - so try to pass just Invoice, without SBD.

**Reply by alfchristensen** [reply]
When removing the SBDH part of the document I got this error message:
`Exception in thread "main" network.oxalis.api.lang.OxalisContentException: Content does not contain SBDH.`

I am trying to respond to the test-bed as that would prove the solution works.

Full log file attached. 
[terminal-log.txt](https://github.com/OxalisCommunity/oxalis/files/11733256/terminal-log.txt)

For the record I can implement the solution using java 11 in my use case, so that is not a game stopper for me. It seems like the signing part is where something is wrong in my code.

Here is the code I am using for the POC:
`package org.example;

import network.oxalis.api.lang.OxalisContentException;
import network.oxalis.api.lang.OxalisTransmissionException;
import network.oxalis.api.outbound.TransmissionRequest;
import network.oxalis.api.outbound.TransmissionResponse;
import network.oxalis.api.outbound.Transmitter;
import network.oxalis.outbound.OxalisOutboundComponent;
import network.oxalis.outbound.transmission.TransmissionRequestBuilder;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Paths;

public class Main {
    public static void main(String[] args) throws IOException, OxalisContentException, OxalisTransmissionException {

        InputStream inputStream = Files.newInputStream(Paths.get("document.xml"));

        OxalisOutboundComponent oxalisOutboundComponent = new OxalisOutboundComponent();
        TransmissionRequestBuilder transmissionRequestBuilder = oxalisOutboundComponent.getTransmissionRequestBuilder();

        // Make request for Oxalis
        transmissionRequestBuilder.payLoad(inputStream);

        // Send request
        TransmissionRequest transmissionRequest = transmissionRequestBuilder.build();
        Transmitter transmitter = oxalisOutboundComponent.getTransmitter();
        TransmissionResponse transmissionResponse = transmitter.transmit(transmissionRequest);

    }
}`

Do you see something wrong with my code? Could it be something with the inputstream, or maybe I should sign it somehow?

**Reply by dladlk** [reply]
I mean that you can use Oxalis Document Sniffer project, which is used in TransmissionRequestBuilder if there is no SBDH:

https://github.com/OxalisCommunity/oxalis/blob/88597d0b1eb46a84425b48f41661b3f2f368c4c7/oxalis-outbound/src/main/java/network/oxalis/outbound/transmission/TransmissionRequestBuilder.java#L218

The issue with signatures is that your SBD is invalid - and now you just miss it at all.

When you sent previous SBDH, it was incorrectly defined - as was mentioned in an unsigned error message from the server.

So you just need to fix your SBDH or generate a new valid one or embed SBD wrapping into your transmission code.

**Reply by alfchristensen** [reply]
That's a bit puzzling. I have a document generated by the testbed and sent to my access point running on Azure as I mentioned earlier. I copied this document, and tried to send the exact same document back. I also tried to swap the sender and receiver as the testbed says. 

Now I have tried the following 4 combinations:
1. Send the exact same document back
2. Swap only sender and receiver and send
3. Remove SBDH and send
4. Remove both SBD and SBDH and send only the Invoice only

I have attached the xml documents and log output with correlating file names. Do you see if there is there something wrong with the document I received from the testbed?

[document.log](https://github.com/OxalisCommunity/oxalis/files/11734917/document.log)
[document_without_sbdh.log](https://github.com/OxalisCommunity/oxalis/files/11734918/document_without_sbdh.log)
[invoice_only.log](https://github.com/OxalisCommunity/oxalis/files/11734919/invoice_only.log)
[xml_documents.zip](https://github.com/OxalisCommunity/oxalis/files/11734932/xml_documents.zip)

**Reply by dladlk** [reply]
WOW! Very interesting :) 

Please try to modify your SBD example like this - avoid any line breaks and white spaces before/after in data, mentioned in Phase4 error message (Phase4 is an implementation of AS4 on the rejecting server whose exception class name we see in response):
![image](https://github.com/OxalisCommunity/oxalis/assets/41484081/bcd23622-e3e4-40fd-9fcf-87af7af897d2)

So it should be like this:

```
<sh:InstanceIdentifier>urn:oasis:names:specification:ubl:schema:xsd:Invoice-2::Invoice##urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0::2.1</sh:InstanceIdentifier>
...
<cbc:CustomizationID>urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0</cbc:CustomizationID>
```

and NOT like it is now:

```
<sh:InstanceIdentifier>
                    urn:oasis:names:specification:ubl:schema:xsd:Invoice-2::Invoice##urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0::2.1</sh:InstanceIdentifier>
                <sh:Identifier>busdox-docid-qns</sh:Identifier>
...
<cbc:CustomizationID>
            urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0</cbc:CustomizationID>
```

I just read the error message you receive from Phase4 instance - but carefully KEEPING ORIGINAL white spaces and line breaks:

```
<eb:Description xml:lang="en">Invoked AS4 message processor SPI com.helger.phase4.peppol.servlet.Phase4PeppolServletMessageProcessorSPI@52ddb896 on 'ad7771e4-3ec6-4bfc-ac76-076a49495614@host.docker.internal' returned a failure: Failed to extract the Peppol data from SBDH. Technical details: com.helger.peppol.sbdh.read.PeppolSBDHDocumentReadException - The 'DocumentIdentification/Standard' element has the illegal value 'urn:oasis:names:specification:ubl:schema:xsd:Invoice-2'. It must equal the namespace URI of the business message ('urn:oasis:names:specification:ubl:schema:xsd:Invoice-2'). It must also be the beginning of the document type identifier value ('
                    urn:oasis:names:specification:ubl:schema:xsd:Invoice-2::Invoice##urn:cen.eu:en16931:2017#compliant#urn:fdc:peppol.eu:2017:poacc:billing:3.0::2.1').</eb:Description>
```

Screenshot to highlight the issue:
![image](https://github.com/OxalisCommunity/oxalis/assets/41484081/9880c970-1f6b-4261-bb0b-6a1210371138)


As we can see in the error message, it says that **document type identifier value** starts with a line break and is prepend with spaces - and I guess that is the reason of the rejection.

Please try to modify the XML as I mentioned - avoid unnecessarily line breaks and white spaces before and after these two validated definitions. I agree that it should work even with these spaces, but it could be that Phase4 validation does not trim them before comparison - and that is why your document is rejected...

Please try it and write back if such "trimming" helped - in this case it can be an important message to Philip, the author of Phase4...

**Reply by alfchristensen** [reply]
This is strange. I had to restart my test in the testbed. I took the same XML as was sent to my access point and tried to send it back. That resulted in the expected error about the sender and recipients not not being correct. Then I carefully only swapped the sender and receiver, and it went straight back to the previous error message. I did this using Notepad ++ with Unix (LF) and UTF-8.

Is it possible that the testbed sends me an XML it won't accept back?

Attached files correlate by name:
[from_testbed.log](https://github.com/OxalisCommunity/oxalis/files/11739830/from_testbed.log)
[to_testbed.log](https://github.com/OxalisCommunity/oxalis/files/11739831/to_testbed.log)
[testbed.zip](https://github.com/OxalisCommunity/oxalis/files/11739861/testbed.zip)

**Reply by dladlk** [reply]
You get the error "Unable to correlate request with Document's Identifier [PNO000567-1-20230608T130800]"

When I look into your XML examples, I see that only SBD headers were changed, I guess, it is also the contents of Invoice AccountingSupplierParty/AccountingCustomerParty should be changed too.

But most important - after you removed the line break in InstanceIdentifier of DOCUMENTID, it fails with another error, so it was the reason of the previous rejections, as now you get another error.

**Reply by alfchristensen** [reply]
Interesting. In the old testbed I had to change it two places like you said. The new testbed states the following:
TC2A.3: AS4 message submissionRunning
The AccessPoint under test is expected to send the document(s) supplied by the TestBed. Sender will now be 0192:810418052, and receiver will be 9922:NGTBCNTRLP1001. No changes should be made to the document and it is essential that the Document ID (Invoice number of the payload) is kept intact.

Is it possible that there is something new or something else going on with the new testbed documents that triggers an error the signing functionality in the org.apache.cxf.ws.policy.PolicyVerificationInInterceptor package?

**Reply by alfchristensen** [reply]
Not sure if this is any help, but here is what ChatGPT-4 responded to the stacktrace:

The error message you're seeing is related to SOAP security policies not being satisfied in your application. The parts of the policy that aren't being satisfied are related to:

The received token not matching the token inclusion requirement ({[http://docs.oasis-open.org/ws-sx/ws-securitypolicy/200702}X509Token](http://docs.oasis-open.org/ws-sx/ws-securitypolicy/200702%7DX509Token)).
The SOAP Body not being signed ({[http://docs.oasis-open.org/ws-sx/ws-securitypolicy/200702}SignedParts](http://docs.oasis-open.org/ws-sx/ws-securitypolicy/200702%7DSignedParts)).
This could be due to a misconfiguration in your security settings, possibly in the way you're signing your SOAP messages, or due to using an incorrect or expired certificate. Here are a few things you could check:

Check your security token: Make sure that you're using the correct security token and it meets the requirements defined by the policy. If you're using an X509 token, check to ensure it's not expired and is included as required by the policy.

Sign your SOAP body: If your SOAP body isn't being signed as required by the policy, you'll need to include this in your message creation process. Check your SOAP message generation code and make sure that it includes a step to sign the body of the message using the correct certificate.

Check your SOAP version: Different SOAP versions may have different requirements for security. Ensure that your application is using the correct SOAP version according to your WS-Security policy.

Policy Misalignment: It is possible that your server and client may have different understandings of what the policy should be. If the client's understanding of the policy (in terms of which parts must be signed, what tokens are needed, etc.) doesn't match what the server expects, these kinds of errors can occur. Make sure to align these policies on both ends.

Check the WSDL file: If you're generating client-side code from a WSDL file, make sure that this file is up to date and matches the server's implementation.

**Reply by dladlk** [reply]
> Is it possible that there is something new or something else going on with the new testbed documents that triggers an error the signing functionality in the org.apache.cxf.ws.policy.PolicyVerificationInInterceptor package?

As I understand from https://github.com/OxalisCommunity/Oxalis-AS4/issues/205 , Phase4 implementation of AS4 Peppol profile does not sign any error responses. That is why Oxalis fails with this error when server rejects a document. And Phase4 is used in this new testbed.

I still hope that Phase4 is updated to sign the rejections as it does not sound logical for me...

So I would recommend you to look into logged original response for now to see what it reports as failure.

For example, in the last case, it could be that the document info is expired, or there is some issue with finding it... So try to get a newest document (and I guess a new document id), do the SBD change and send it back as fast as possible.

In general, new testbed is released just a week ago, so theoretically there could be issues with it.

**Reply by alfchristensen** [reply]
> For example, in the last case, it could be that the document info is expired, or there is some issue with finding it... So try to get a newest document (and I guess a new document id), do the SBD change and send it back as fast as possible.

Just reset the testbed and received a new document. Changed only the sender and receiver as specified and tried to respond. Still the same error of soap message not being signed:

`Jun 14, 2023 2:50:11 PM org.apache.cxf.ws.policy.PolicyVerificationInInterceptor handle
SEVERE: Inbound policy verification failed: These policy alternatives can not be satisfied: 
{http://docs.oasis-open.org/ws-sx/ws-securitypolicy/200702}X509Token: The received token does not match the token inclusion requirement
{http://docs.oasis-open.org/ws-sx/ws-securitypolicy/200702}SignedParts: Soap Body is not SIGNED`

> So I would recommend you to look into logged original response for now to see what it reports as failure.

Judging from the logs it does not seem like Oxalis tries to send at all? Could that be the case, or do you think it is just the response from the testbed that is causing this? Referring to what you said in your previous post. The testbed is running but shows nothing in the Audit log (attached picture). 

One thing: I have the oxalis:latest-11 (docker image running 6.0.0-RC4) running in an Azure container. It passes the certificate test and receives the document. THEN I copy the file to my application utilizing the maven dependency running on my local computer, and try to send it back. Using the same p12 certificate as the docker container and the same jks file. 

Do you think I should contact the people working with the testbed, or du you think this error relates to my implementation or files?

logfile from my application and xml attached.

[new_trial_06_14.zip](https://github.com/OxalisCommunity/oxalis/files/11745952/new_trial_06_14.zip)
[new_trial_06_14.log](https://github.com/OxalisCommunity/oxalis/files/11745955/new_trial_06_14.log)
![testbed_status](https://github.com/OxalisCommunity/oxalis/assets/29151576/3ab338ab-8c15-4981-99f1-29b08bff02b9)

**Reply by dladlk** [reply]
> Judging from the logs it does not seem like Oxalis tries to send at all?
In your logs you can see, that Oxalis made a request:

```
2023-06-14 14:54:54,594 INFO [org.apache.cxf.services.DummyImpl.REQ_OUT] REQ_OUT
    Address: https://phase4-controller.testbed.peppol.org/as4
    HttpMethod: POST
    Content-Type: multipart/related; type="application/soap+xml"; boundary="uuid:832275bb-ce69-4bd8-9205-892ac75a50f0"; start="<root.message@cxf.apache.org>"; start-info="application/soap+xml"
    ExchangeId: 398a5b96-4859-4d19-85bc-4e624074fe0c
    ServiceName: outbound-service
    PortName: port
    PortTypeName: DummyImpl
    Headers: {Accept=*/*, User-Agent=Oxalis 6.0.0-RC3 / AS4 6.0.0-RC3 / PNO000567}
    Payload: 
...
```

And in 400 ms it received response:
```
2023-06-14 14:54:54,931 INFO [org.apache.cxf.services.DummyImpl.RESP_IN] RESP_IN
    Address: https://phase4-controller.testbed.peppol.org/as4
    Content-Type: application/soap+xml;charset=UTF-8
    ResponseCode: 200
    ExchangeId: 398a5b96-4859-4d19-85bc-4e624074fe0c
    ServiceName: outbound-service
    PortName: port
...
```

You can see the response after `Payload:` text. If you copy the contents and format it as XML, you can see the error message inside Error\Description tag:

```
<?xml version="1.0" encoding="UTF-8"?>
<S12:Envelope xmlns:S12="http://www.w3.org/2003/05/soap-envelope">
	<S12:Header>
		<eb:Messaging...>
			<eb:SignalMessage>
				<eb:MessageInfo>
					<eb:Timestamp>2023-06-14T12:54:53.311Z</eb:Timestamp>
					<eb:MessageId>da8983b6-9259-4d10-a2f3-2fca7bcd6c08@phase4.openpeppol.playground</eb:MessageId>
					<eb:RefToMessageId>e06e59c4-060e-4907-b623-1d050d164d59@host.docker.internal</eb:RefToMessageId>
				</eb:MessageInfo>
				<eb:Error category="Content" errorCode="EBMS:0004" refToMessageInError="e06e59c4-060e-4907-b623-1d050d164d59@host.docker.internal" severity="failure" shortDescription="Other">
					<eb:Description xml:lang="en">Invoked AS4 message processor SPI com.helger.phase4.peppol.servlet.Phase4PeppolServletMessageProcessorSPI@2244b9b5 on 'e06e59c4-060e-4907-b623-1d050d164d59@host.docker.internal' returned a failure: TBErrorResponse{errorCode='-4115', errorDescription='Unable to correlate request with Document's Identifier [PNO000567-4-20230614T094641]', trace='d068d330-be90-4726-9f27-339b6e060f75'}</eb:Description>
					<eb:ErrorDetail>An undefined error occurred.</eb:ErrorDetail>
				</eb:Error>
			</eb:SignalMessage>
		</eb:Messaging>
	</S12:Header>
	<S12:Body/>
</S12:Envelope>
```

which is again similar as before.

Error message:

> Invoked AS4 message processor SPI com.helger.phase4.peppol.servlet.Phase4PeppolServletMessageProcessorSPI@2244b9b5 on 'e06e59c4-060e-4907-b623-1d050d164d59@host.docker.internal' returned a failure: TBErrorResponse{errorCode='-4115', errorDescription='**Unable to correlate request with Document's Identifier [PNO000567-4-20230614T094641]**', trace='d068d330-be90-4726-9f27-339b6e060f75'}

But Oxalis expects even errors to be signed - that is why you get another exception further.
Anyway, I guess it worth contacting Peppol testbed support and send them the whole above error message including trace ID, maybe there is some issue now...

**Reply by alfchristensen** [reply]
Hello again @dladlk 

Just got a response on my Jira ticket from Phillip Helger. He stated that this issue is caused by the requirement for signed error responses in Oxalis, and pointed me this thread: https://github.com/OxalisCommunity/Oxalis-AS4/issues/205

After having read the discussion, I see there are some disagreement on policy and thus implementation. Do you have any suggestions for what I can do to get around this in the meantime? If I you will consider changing the Oxalis implementation, that would be very good for me to know also.

Best Regards
Alf Christensen
